File: controller/routes/lifecycle.py
"""Model lifecycle endpoints: recipes CRUD, launch, evict."""

from __future__ import annotations

import asyncio
import logging
import time
from pathlib import Path
from typing import Optional

import httpx
import psutil
from fastapi import APIRouter, Depends, HTTPException

from ..config import settings
from ..events import event_manager
from ..models import LaunchResult, Recipe
from ..process import evict_model, find_inference_process, kill_process, launch_model
from ..store import RecipeStore
from .system import get_launching_recipe_id, set_launching_recipe_id

router = APIRouter()
logger = logging.getLogger(__name__)

_switch_lock = asyncio.Lock()
_launch_cancel_events: dict[str, asyncio.Event] = {}


def get_store() -> RecipeStore:
    """Get recipe store instance."""
    from ..app import get_store as _get_store
    return _get_store()


# --- Recipes ---
@router.get("/recipes", tags=["Recipes"])
async def list_recipes(store: RecipeStore = Depends(get_store)):
    """List all recipes."""
    recipes = store.list()
    current = find_inference_process(settings.inference_port)
    result = []
    launching_id = get_launching_recipe_id()
    for r in recipes:
        status = "stopped"
        if launching_id == r.id:
            status = "starting"
        if current:
            if current.served_model_name and r.served_model_name == current.served_model_name:
                status = "running"
            elif current.model_path:
                if r.model_path in current.model_path or current.model_path in r.model_path:
                    status = "running"
                elif current.model_path.split("/")[-1] == r.model_path.split("/")[-1]:
                    status = "running"
        result.append({**r.model_dump(), "status": status})
    return result


@router.get("/recipes/{recipe_id}", tags=["Recipes"])
async def get_recipe(recipe_id: str, store: RecipeStore = Depends(get_store)):
    """Get a recipe by ID."""
    recipe = store.get(recipe_id)
    if not recipe:
        raise HTTPException(status_code=404, detail="Recipe not found")
    return recipe


@router.post("/recipes", tags=["Recipes"])
async def create_recipe(recipe: Recipe, store: RecipeStore = Depends(get_store)):
    """Create or update a recipe."""
    store.save(recipe)
    return {"success": True, "id": recipe.id}


@router.put("/recipes/{recipe_id}", tags=["Recipes"])
async def update_recipe(recipe_id: str, recipe: Recipe, store: RecipeStore = Depends(get_store)):
    """Update a recipe by ID."""
    if recipe.id != recipe_id:
        recipe.id = recipe_id
    store.save(recipe)
    return {"success": True, "id": recipe.id}


@router.delete("/recipes/{recipe_id}", tags=["Recipes"])
async def delete_recipe(recipe_id: str, store: RecipeStore = Depends(get_store)):
    """Delete a recipe."""
    if not store.delete(recipe_id):
        raise HTTPException(status_code=404, detail="Recipe not found")
    return {"success": True}


# --- Model lifecycle ---
@router.post("/launch/{recipe_id}", response_model=LaunchResult, tags=["Lifecycle"])
async def launch(recipe_id: str, force: bool = False, store: RecipeStore = Depends(get_store)):
    """Launch a model by recipe ID with real-time progress updates."""
    global _launch_cancel_events

    recipe = store.get(recipe_id)
    if not recipe:
        raise HTTPException(status_code=404, detail="Recipe not found")

    # Check if another launch is in progress and preempt it
    current_launching = get_launching_recipe_id()
    if current_launching and current_launching != recipe_id:
        preempted_recipe_id = current_launching
        logger.info(f"Preempting launch of {preempted_recipe_id} for {recipe_id}")
        await event_manager.publish_launch_progress(
            recipe_id, "preempting", f"Cancelling {preempted_recipe_id}...", progress=0.0
        )
        await event_manager.publish_launch_progress(
            preempted_recipe_id, "cancelled", f"Preempted by {recipe_id}", progress=0.0
        )

        cancel_event = _launch_cancel_events.get(preempted_recipe_id)
        if cancel_event:
            cancel_event.set()

        await evict_model(force=True)
        await asyncio.sleep(1)

    cancel_event = asyncio.Event()
    _launch_cancel_events[recipe_id] = cancel_event
    set_launching_recipe_id(recipe_id)

    try:
        try:
            await asyncio.wait_for(_switch_lock.acquire(), timeout=2.0)
        except asyncio.TimeoutError:
            logger.info(f"Lock contention - force preempting for {recipe_id}")
            await evict_model(force=True)
            await asyncio.sleep(1)
            await _switch_lock.acquire()

        try:
            await event_manager.publish_launch_progress(
                recipe_id, "evicting", "Clearing VRAM...", progress=0.0
            )
            await evict_model(force=True)
            await asyncio.sleep(1)

            if cancel_event.is_set():
                await event_manager.publish_launch_progress(
                    recipe_id, "cancelled", "Preempted by another launch", progress=0.0
                )
                return LaunchResult(success=False, pid=None, message="Launch cancelled", log_file=None)

            await event_manager.publish_launch_progress(
                recipe_id, "launching", f"Starting {recipe.name}...", progress=0.25
            )

            success, pid, message = await launch_model(recipe)

            if not success:
                await event_manager.publish_launch_progress(
                    recipe_id, "error", message, progress=0.0
                )
                return LaunchResult(success=False, pid=None, message=message, log_file=None)

            await event_manager.publish_launch_progress(
                recipe_id, "waiting", "Waiting for model to load...", progress=0.5
            )

            start = time.time()
            timeout = 300
            ready = False
            crashed = False

            while time.time() - start < timeout:
                if cancel_event.is_set():
                    logger.info(f"Launch of {recipe_id} cancelled during load")
                    await event_manager.publish_launch_progress(
                        recipe_id, "cancelled", "Preempted by another launch", progress=0.0
                    )
                    if pid:
                        try:
                            proc = psutil.Process(pid)
                            for child in proc.children(recursive=True):
                                child.kill()
                            proc.kill()
                        except psutil.NoSuchProcess:
                            pass
                    return LaunchResult(success=False, pid=None, message="Launch cancelled", log_file=None)

                try:
                    if pid and not psutil.pid_exists(pid):
                        crashed = True
                        break
                except Exception:
                    pass

                try:
                    async with httpx.AsyncClient(timeout=5) as client:
                        r = await client.get(f"http://localhost:{settings.inference_port}/health")
                        if r.status_code == 200:
                            ready = True
                            break
                except Exception:
                    pass

                elapsed = int(time.time() - start)
                await event_manager.publish_launch_progress(
                    recipe_id, "waiting",
                    f"Loading model... ({elapsed}s)",
                    progress=0.5 + (elapsed / timeout) * 0.5
                )
                await asyncio.sleep(2)

            if crashed:
                log_file = Path(f"/tmp/vllm_{recipe_id}.log")
                error_tail = ""
                if log_file.exists():
                    try:
                        error_tail = log_file.read_text()[-1000:]
                    except Exception:
                        pass
                await event_manager.publish_launch_progress(
                    recipe_id, "error", "Model process crashed. Check logs for details.", progress=0.0
                )
                return LaunchResult(success=False, pid=None, message=f"Process crashed: {error_tail[-200:]}", log_file=str(log_file))

            if ready:
                await event_manager.publish_launch_progress(
                    recipe_id, "ready", "Model is ready!", progress=1.0
                )
                logger.info(f"Model {recipe_id} is ready")
                return LaunchResult(
                    success=True,
                    pid=pid,
                    message="Model is ready",
                    log_file=f"/tmp/vllm_{recipe_id}.log",
                )
            else:
                try:
                    if pid:
                        await kill_process(pid, force=True)
                except Exception:
                    pass

                log_file = Path(f"/tmp/vllm_{recipe_id}.log")
                error_tail = ""
                if log_file.exists():
                    try:
                        error_tail = log_file.read_text()[-1000:]
                    except Exception:
                        pass
                await event_manager.publish_launch_progress(
                    recipe_id, "error", "Model failed to become ready (timeout)", progress=0.0
                )
                return LaunchResult(
                    success=False,
                    pid=None,
                    message=f"Model failed to become ready (timeout): {error_tail[-200:]}",
                    log_file=str(log_file),
                )
        finally:
            _switch_lock.release()
    finally:
        if get_launching_recipe_id() == recipe_id:
            set_launching_recipe_id(None)
        if _launch_cancel_events.get(recipe_id) is cancel_event:
            del _launch_cancel_events[recipe_id]


@router.post("/evict", tags=["Lifecycle"])
async def evict(force: bool = False):
    """Stop the running model."""
    async with _switch_lock:
        pid = await evict_model(force=force)
    logger.info(f"Model evicted (pid={pid})")
    return {"success": True, "evicted_pid": pid}


@router.get("/wait-ready", tags=["Lifecycle"])
async def wait_ready(timeout: int = 300):
    """Wait for inference backend to be ready."""
    start = time.time()
    while time.time() - start < timeout:
        try:
            async with httpx.AsyncClient(timeout=5) as client:
                r = await client.get(f"http://localhost:{settings.inference_port}/health")
                if r.status_code == 200:
                    return {"ready": True, "elapsed": int(time.time() - start)}
        except Exception:
            pass
        await asyncio.sleep(2)

    return {"ready": False, "elapsed": timeout, "error": "Timeout waiting for backend"}

File: controller/src/routes/lifecycle.ts
import type { Hono } from "hono";
import { existsSync, readFileSync } from "node:fs";
import { join } from "node:path";
import type { AppContext } from "../types/context";
import type { LaunchResult, Recipe } from "../types/models";
import { AsyncLock, delay } from "../core/async";
import { badRequest, notFound } from "../core/errors";
import { parseRecipe } from "../stores/recipe-serializer";

const switchLock = new AsyncLock();
const launchCancelControllers = new Map<string, AbortController>();

/**
 * Register lifecycle routes.
 * @param app - Hono app.
 * @param context - App context.
 */
export const registerLifecycleRoutes = (app: Hono, context: AppContext): void => {
  /**
   * Check if a process id exists.
   * @param pid - Process id.
   * @returns True if process exists.
   */
  const pidExists = (pid: number): boolean => {
    try {
      process.kill(pid, 0);
      return true;
    } catch {
      return false;
    }
  };

  /**
   * Read last N characters from a log file.
   * @param path - Log file path.
   * @param limit - Max chars.
   * @returns Log tail string.
   */
  const readLogTail = (path: string, limit: number): string => {
    if (!existsSync(path)) {
      return "";
    }
    try {
      const content = readFileSync(path, "utf-8");
      return content.slice(Math.max(0, content.length - limit));
    } catch {
      return "";
    }
  };

  const serializeRecipeDetail = (recipe: Recipe): Record<string, unknown> => {
    const payload: Record<string, unknown> = {
      ...recipe,
      tp: recipe.tensor_parallel_size,
      pp: recipe.pipeline_parallel_size,
    };
    delete payload["tensor_parallel_size"];
    delete payload["pipeline_parallel_size"];
    return payload;
  };

  app.get("/recipes", async (ctx) => {
    const recipes = context.stores.recipeStore.list();
    const current = await context.processManager.findInferenceProcess(context.config.inference_port);
    const launchingId = context.launchState.getLaunchingRecipeId();
    const result = recipes.map((recipe) => {
      let status = "stopped";
      if (launchingId === recipe.id) {
        status = "starting";
      }
      if (current) {
        if (current.served_model_name && recipe.served_model_name === current.served_model_name) {
          status = "running";
        } else if (current.model_path) {
          // Compare normalized paths (exact match)
          const normalize = (p: string): string => p.replace(/\/+$/, "");
          if (normalize(recipe.model_path) === normalize(current.model_path)) {
            status = "running";
          } else if (current.model_path.split("/").pop() === recipe.model_path.split("/").pop()) {
            status = "running";
          }
        }
      }
      return { ...recipe, status };
    });
    return ctx.json(result);
  });

  app.get("/recipes/:recipeId", async (ctx) => {
    const recipeId = ctx.req.param("recipeId");
    const recipe = context.stores.recipeStore.get(recipeId);
    if (!recipe) {
      throw notFound("Recipe not found");
    }
    return ctx.json(serializeRecipeDetail(recipe));
  });

  app.post("/recipes", async (ctx) => {
    const body = await ctx.req.json();
    try {
      const recipe = parseRecipe(body);
      context.stores.recipeStore.save(recipe);
      return ctx.json({ success: true, id: recipe.id });
    } catch (error) {
      throw badRequest(String(error));
    }
  });

  app.put("/recipes/:recipeId", async (ctx) => {
    const recipeId = ctx.req.param("recipeId");
    const body = await ctx.req.json();
    try {
      const recipe = parseRecipe({ ...body, id: recipeId });
      context.stores.recipeStore.save(recipe);
      return ctx.json({ success: true, id: recipe.id });
    } catch (error) {
      throw badRequest(String(error));
    }
  });

  app.delete("/recipes/:recipeId", async (ctx) => {
    const recipeId = ctx.req.param("recipeId");
    const deleted = context.stores.recipeStore.delete(recipeId);
    if (!deleted) {
      throw notFound("Recipe not found");
    }
    return ctx.json({ success: true });
  });

  app.post("/launch/:recipeId", async (ctx) => {
    const recipeId = ctx.req.param("recipeId");
    const recipe = context.stores.recipeStore.get(recipeId);
    if (!recipe) {
      throw notFound("Recipe not found");
    }

    const currentLaunching = context.launchState.getLaunchingRecipeId();
    if (currentLaunching && currentLaunching !== recipeId) {
      await context.eventManager.publishLaunchProgress(recipeId, "preempting", `Cancelling ${currentLaunching}...`, 0);
      await context.eventManager.publishLaunchProgress(currentLaunching, "cancelled", `Preempted by ${recipeId}`, 0);
      const cancel = launchCancelControllers.get(currentLaunching);
      if (cancel) {
        cancel.abort();
      }
      await context.processManager.evictModel(true);
      await delay(1000);
    }

    const cancelController = new AbortController();
    launchCancelControllers.set(recipeId, cancelController);
    context.launchState.setLaunchingRecipeId(recipeId);

    let releaseLock: (() => void) | null = null;
    try {
      releaseLock = await switchLock.acquireWithTimeout(2000);
      if (!releaseLock) {
        await context.processManager.evictModel(true);
        await delay(1000);
        releaseLock = await switchLock.acquire();
      }

      await context.eventManager.publishLaunchProgress(recipeId, "evicting", "Clearing VRAM...", 0);
      await context.processManager.evictModel(true);
      await delay(1000);

      if (cancelController.signal.aborted) {
        await context.eventManager.publishLaunchProgress(recipeId, "cancelled", "Preempted by another launch", 0);
        return ctx.json({ success: false, pid: null, message: "Launch cancelled", log_file: null } as LaunchResult);
      }

      await context.eventManager.publishLaunchProgress(recipeId, "launching", `Starting ${recipe.name}...`, 0.25);
      const launch = await context.processManager.launchModel(recipe);
      if (!launch.success) {
        await context.eventManager.publishLaunchProgress(recipeId, "error", launch.message, 0);
        return ctx.json({ success: false, pid: null, message: launch.message, log_file: null } as LaunchResult);
      }

      await context.eventManager.publishLaunchProgress(recipeId, "waiting", "Waiting for model to load...", 0.5);

      const start = Date.now();
      const timeout = 300_000;
      let ready = false;
      let fatalError: string | null = null;
      const fatalPatterns = [
        "raise ValueError",
        "raise RuntimeError",
        "CUDA out of memory",
        "OutOfMemoryError",
        "torch.OutOfMemoryError",
        "not enough memory",
        "Cannot allocate",
        "larger than the available KV cache memory",
        "EngineCore failed to start",
      ];

      const logFilePath = join("/tmp", `vllm_${recipeId}.log`);

      while (Date.now() - start < timeout) {
        if (cancelController.signal.aborted) {
          await context.eventManager.publishLaunchProgress(recipeId, "cancelled", "Preempted by another launch", 0);
          if (launch.pid && pidExists(launch.pid)) {
            await context.processManager.killProcess(launch.pid, true);
          }
          return ctx.json({ success: false, pid: null, message: "Launch cancelled", log_file: null } as LaunchResult);
        }

        const logTail = readLogTail(logFilePath, 3000);
        if (logTail) {
          for (const pattern of fatalPatterns) {
            if (logTail.includes(pattern)) {
              const lines = logTail.split("\n");
              const index = lines.findIndex((line) => line.includes(pattern));
              if (index >= 0) {
                fatalError = lines.slice(Math.max(0, index - 1), index + 3).join("\n");
              }
              break;
            }
          }
        }
        if (fatalError) {
          break;
        }

        try {
          const controller = new AbortController();
          const timeoutHandle = setTimeout(() => controller.abort(), 5000);
          const response = await fetch(`http://localhost:${context.config.inference_port}/health`, {
            signal: controller.signal,
          });
          clearTimeout(timeoutHandle);
          if (response.status === 200) {
            ready = true;
            break;
          }
        } catch {
          ready = false;
        }

        const elapsed = Math.floor((Date.now() - start) / 1000);
        await context.eventManager.publishLaunchProgress(
          recipeId,
          "waiting",
          `Loading model... (${elapsed}s)`,
          0.5 + (elapsed / (timeout / 1000)) * 0.5,
        );
        await delay(2000);
      }

      if (fatalError) {
        if (launch.pid) {
          await context.processManager.killProcess(launch.pid, true);
        }
        await context.eventManager.publishLaunchProgress(recipeId, "error", `Fatal error detected: ${fatalError.slice(0, 100)}`, 0);
        return ctx.json({
          success: false,
          pid: null,
          message: `Fatal error: ${fatalError.slice(0, 300)}`,
          log_file: logFilePath,
        } as LaunchResult);
      }

      if (ready) {
        await context.eventManager.publishLaunchProgress(recipeId, "ready", "Model is ready!", 1.0);
        return ctx.json({
          success: true,
          pid: launch.pid ?? null,
          message: "Model is ready",
          log_file: logFilePath,
        } as LaunchResult);
      }

      if (launch.pid) {
        await context.processManager.killProcess(launch.pid, true);
      }
      const errorTail = readLogTail(logFilePath, 1000);
      await context.eventManager.publishLaunchProgress(recipeId, "error", "Model failed to become ready (timeout)", 0);
      return ctx.json({
        success: false,
        pid: null,
        message: `Model failed to become ready (timeout): ${errorTail.slice(-200)}`,
        log_file: logFilePath,
      } as LaunchResult);
    } finally {
      if (releaseLock) {
        releaseLock();
      }
      if (context.launchState.getLaunchingRecipeId() === recipeId) {
        context.launchState.setLaunchingRecipeId(null);
      }
      const controller = launchCancelControllers.get(recipeId);
      if (controller === cancelController) {
        launchCancelControllers.delete(recipeId);
      }
    }
  });

  app.post("/launch/:recipeId/cancel", async (ctx) => {
    const recipeId = ctx.req.param("recipeId");
    const cancel = launchCancelControllers.get(recipeId);
    if (!cancel) {
      const current = context.launchState.getLaunchingRecipeId();
      if (current !== recipeId) {
        throw notFound(`No launch in progress for ${recipeId}`);
      }
      await context.processManager.evictModel(true);
      return ctx.json({ success: true, message: "Launch aborted via eviction" });
    }
    cancel.abort();
    await context.processManager.evictModel(true);
    return ctx.json({ success: true, message: `Launch of ${recipeId} cancelled` });
  });

  app.post("/evict", async (ctx) => {
    const release = await switchLock.acquire();
    try {
      const pid = await context.processManager.evictModel(Boolean(ctx.req.query("force")));
      return ctx.json({ success: true, evicted_pid: pid });
    } finally {
      release();
    }
  });

  app.get("/wait-ready", async (ctx) => {
    const timeout = Number(ctx.req.query("timeout") ?? 300);
    const start = Date.now();
    while (Date.now() - start < timeout * 1000) {
      try {
        const controller = new AbortController();
        const timeoutHandle = setTimeout(() => controller.abort(), 5000);
        const response = await fetch(`http://localhost:${context.config.inference_port}/health`, {
          signal: controller.signal,
        });
        clearTimeout(timeoutHandle);
        if (response.status === 200) {
          return ctx.json({ ready: true, elapsed: Math.floor((Date.now() - start) / 1000) });
        }
      } catch {
        await delay(2000);
      }
      await delay(2000);
    }
    return ctx.json({ ready: false, elapsed: timeout, error: "Timeout waiting for backend" });
  });
};
